File: Appointment/AppointmentDetailViewController.m
#import "AppointmentDetailViewController.h"
#import "MapViewController.h"

#import "Appointment-Swift.h"

@implementation AppointmentDetailViewController

@synthesize appointment;

// MARK: - Lifecycle

- (void)viewDidLoad {
	[super viewDidLoad];
	
	self.tableView.allowsSelection = NO;

    [self setUpMenu];

    [self.tableView registerClass:AppointmentDetailTableViewCell.class forCellReuseIdentifier:AppointmentDetailTableViewCell.reuseIdentifier];
	
	self.tableView.rowHeight = UITableViewAutomaticDimension;
	self.tableView.estimatedRowHeight = 50.0;
	
	self.title = appointment.clientName;
}

// MARK: - Menu Actions

- (void)setUpMenu {
    UIAction *phoneAction = [UIAction actionWithTitle:@"Call client" image:[UIImage systemImageNamed:@"phone"] identifier:nil handler:^(__kindof UIAction * _Nonnull action) {
        [self callContact];
    }];
    
    UIAction *emailAction = [UIAction actionWithTitle:@"Email client" image:[UIImage systemImageNamed:@"envelope"] identifier:nil handler:^(__kindof UIAction * _Nonnull action) {
        [self showEmailComposer];
    }];

    UIAction *mapAction = [UIAction actionWithTitle:@"Find on map" image:[UIImage systemImageNamed:@"mappin.and.ellipse"] identifier:nil handler:^(__kindof UIAction * _Nonnull action) {
        [self goToMapView];
    }];

    UIAction *addContactAction = [UIAction actionWithTitle:@"Add contact" image:[UIImage systemImageNamed:@"person.crop.circle.badge.plus"] identifier:nil handler:^(__kindof UIAction * _Nonnull action) {
        [self createNewContact];
    }];

    UIAction *calendarAction = [UIAction actionWithTitle:@"Calendar" image:[UIImage systemImageNamed:@"calendar.badge.plus"] identifier:nil handler:^(__kindof UIAction * _Nonnull action) {
        [self createNewCalendarEvent];
    }];

    UIMenu *menu = [UIMenu menuWithTitle:@"" children:@[ phoneAction, emailAction, mapAction, addContactAction, calendarAction]];

    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage systemImageNamed:@"ellipsis.circle"] menu:menu];
}

- (void)callContact {
	NSURL *phoneURL = [NSURL URLWithString:[NSString stringWithFormat: @"telprompt:%@", appointment.clientPhone]];
	
	if ([[UIApplication sharedApplication] canOpenURL:phoneURL]) {
		[[UIApplication sharedApplication] openURL:phoneURL options:@{} completionHandler:nil];
	} else {
		UIAlertController *alertController = [UIAlertController
											  alertControllerWithTitle:@"Error"
											  message:@"Your device cannot make calls."
											  preferredStyle:UIAlertControllerStyleAlert];
		
		UIAlertAction *dismissAction = [UIAlertAction
										actionWithTitle:@"Dismiss"
										style:UIAlertActionStyleCancel
										handler:nil];
		
		[alertController addAction:dismissAction];
		
		[self presentViewController:alertController animated:YES completion:nil];
	}
}

- (void)showEmailComposer {
	if ([MFMailComposeViewController canSendMail]) {
		MFMailComposeViewController *controller = [[MFMailComposeViewController alloc] init];
		
		controller.navigationBar.tintColor = [UIColor whiteColor];
		
		controller.mailComposeDelegate = self;
		
		[controller setSubject: [NSString stringWithFormat: @"%@ Appointment With %@", appointment.appointmentDateString, appointment.clientName]];
		[controller setToRecipients: @[appointment.clientEmail]];
		[controller setMessageBody:appointment.emailBodyString isHTML:YES];
		
		[self presentViewController:controller animated:YES completion:nil];
	} else {
		UIAlertController *alertController = [UIAlertController
											  alertControllerWithTitle:@"Failure"
											  message:@"Your device doesn't support the composer sheet"
											  preferredStyle:UIAlertControllerStyleAlert];
		
		UIAlertAction *cancelAction = [UIAlertAction
									   actionWithTitle:@"OK"
									   style:UIAlertActionStyleCancel
									   handler:nil];
		
		[alertController addAction:cancelAction];
		
		[self presentViewController:alertController animated:YES completion:nil];
	}
}

- (void)goToMapView {
	MapViewController *mapVC = [[MapViewController alloc] init];
	mapVC.addressString = [NSString stringWithFormat:@"%@ %@", appointment.address, appointment.zipCode];
	
	[self.navigationController pushViewController:mapVC animated:YES];
}

- (void)createNewContact {
	[ContactHandler createNewContactWith:appointment on:self];
}

- (void)createNewCalendarEvent {
	[CalendarEventHandler createNewCalendarEventWith:appointment on:self];
}

// MARK: - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
	return appointment.appointmentProperties.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
	AppointmentDetailTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:AppointmentDetailTableViewCell.reuseIdentifier];

    [cell styleWith:appointment index:indexPath.row];

	return cell;
}

// MARK: - MFMailComposeViewControllerDelegate

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error {
	[self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Appointment/AppointmentDetailViewController.swift
//
//  AppointmentDetailViewController.swift
//  Appointment
//
//  Created by Jeffrey Moran on 10/15/20.
//  Copyright © 2020 Jeff Moran. All rights reserved.
//

import MessageUI
import UIKit

class AppointmentDetailViewController: UITableViewController {

    // MARK: - Private Properties

    private var appointment: Appointment

    private var phoneUrl: URL? {
        let phoneNumber = (appointment.clientPhone ?? "").filter("0123456789.".contains)

        return URL(string: "telprompt:\(phoneNumber)")
    }

    private var shouldShowPhoneAction: Bool {
        guard let phoneUrl = phoneUrl else { return false }

        return UIApplication.shared.canOpenURL(phoneUrl)
    }

    // MARK: - Initializers

    @objc init(with appointment: Appointment) {
        self.appointment = appointment

        super.init(style: .grouped)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        title = appointment.clientName

        tableView.allowsSelection = false
        tableView.register(AppointmentDetailTableViewCell.self, forCellReuseIdentifier: AppointmentDetailTableViewCell.reuseIdentifier)

        setUpMenu()
    }

    // MARK: - Private Methods

    private func setUpMenu() {
        let phoneAction = UIAction(title: "Call Client", image: UIImage(systemName: "phone")) { _ in
            self.callContact()
        }

        let emailAction = UIAction(title: "Email client", image: UIImage(systemName: "envelope")) { _ in
            self.showEmailComposer()
        }

        let mapAction = UIAction(title: "Find on map", image: UIImage(systemName: "mappin.and.ellipse")) { _ in
            self.goToMapView()
        }

        let addContactAction = UIAction(title: "Add contact", image: UIImage(systemName: "person.crop.circle.badge.plus")) { _ in
            self.createNewContact()
        }

        let calendarAction = UIAction(title: "Add to Calendar", image: UIImage(systemName: "calendar.badge.plus")) { _ in
            self.createNewCalendarEvent()
        }

        var actions = [UIAction]()

        if shouldShowPhoneAction {
            actions.append(phoneAction)
        }

        if MFMailComposeViewController.canSendMail() {
            actions.append(emailAction)
        }

        actions.append(contentsOf: [mapAction, addContactAction, calendarAction])

        let menu = UIMenu(children: actions)

        navigationItem.rightBarButtonItem = UIBarButtonItem(image: UIImage(systemName: "ellipsis.circle"), menu: menu)
    }

    private func callContact() {
        guard let phoneUrl = phoneUrl else { return }

        UIApplication.shared.open(phoneUrl)
    }

    private func showEmailComposer() {
        let controller: MFMailComposeViewController! = MFMailComposeViewController()

        controller.navigationBar.tintColor = .white

        controller.mailComposeDelegate = self

        let subject = "\(appointment.appointmentDateString ?? "") Appointment with \(appointment.clientName ?? "")"
        controller.setSubject(subject)

        controller.setToRecipients([appointment.clientEmail ?? ""])

        controller.setMessageBody(appointment.emailBodyString, isHTML: true)

        present(controller, animated: true)
    }

    private func goToMapView() {
        let addressString = "\(appointment.address ?? "") \(appointment.zipCode ?? "")"

        let mapViewController = MapViewController(with: addressString)

        navigationController?.pushViewController(mapViewController, animated: true)
    }

    private func createNewContact() {
        ContactHandler.createNewContact(with: appointment, on: self)
    }

    private func createNewCalendarEvent() {
        CalendarEventHandler.createNewCalendarEvent(with: appointment, on: self)
    }
}

// MARK: - UITableViewDataSource

extension AppointmentDetailViewController {
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return appointment.appointmentProperties.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: AppointmentDetailTableViewCell.reuseIdentifier) as? AppointmentDetailTableViewCell else {
            fatalError("Wrong cell type!")
        }

        cell.style(with: appointment, index: indexPath.row)

        return cell
    }
}

// MARK: - MFMailComposeViewControllerDelegate

extension AppointmentDetailViewController: MFMailComposeViewControllerDelegate {
    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        controller.dismiss(animated: true)
    }
}
